### PR TITLE
Add pure component transform

### PR DESCRIPTION
**Pull request by thejameskyle**
This is all I have time for now so I'm throwing it up into a PR.

~~Trying to figure out why Jest won't run.~~

Re: https://twitter.com/cpojer/status/656602748050247680

### CODE REVIEW

**Comment by zpao** [comment]
FWIW: tests look like they are running for me, but there are a few pre-existing failures

**Comment by zpao** [comment]
Oh actually I bet it isn't wouldn't be running - add your case in https://github.com/reactjs/react-codemod/blob/master/test/__tests__/transform-tests.js. That won't solve general issues with jest not running but will make sure your test does run when you work that out.

**Comment by cpojer** [comment]
yeah I'll take a look soon-ish at why the tests are failing.

@thejameskyle awesome work! :)

File: transforms/pure-component.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    [j.identifier('props')],
+    body
+  );
+}
+
+module.exports = function(file, api) {
+  const j = api.jscodeshift;
+  const f = j(file.source);
+
+  return j(file.source)
```
Comment (cpojer) [inline transforms/pure-component.js]: This should do a `didTransform` kinda thing and only return a non-null value if the `.size()` of this collection is `> 0`.

See https://github.com/reactjs/react-codemod/blob/master/transforms/pure-render-mixin.js#L172

File: transforms/pure-component.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  return j.functionDeclaration(
+    j.identifier(name),
+    [j.identifier('props')],
+    body
+  );
+}
+
+module.exports = function(file, api) {
+  const j = api.jscodeshift;
+  const f = j(file.source);
```
Comment (cpojer) [inline transforms/pure-component.js]: please copy pasta this: https://github.com/reactjs/react-codemod/blob/master/transforms/pure-render-mixin.js#L19

File: transforms/pure-component.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  return j(file.source)
+    .find(j.ClassDeclaration, REACT_COMPONENT_CLASS_DEFINITION)
```
Comment (cpojer) [inline transforms/pure-component.js]: please filter that the class only has one method called `render` so it skips over more complex cases.

File: transforms/pure-component.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  object: {
+    type: 'ThisExpression'
+  },
+  property: {
+    name: 'props'
+  }
+};
+
+function replaceThisProps(node) {
+  return j(node)
```
Comment (cpojer) [inline transforms/pure-component.js]: stupid question but did you test this? :P `j` is only part of the exported function below and not a global.

File: test/pure-component-test.output.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+var React = require('React');
+
+function render() {
+  return <div/>;
+}
+
+function A(props) {
```
Comment (cpojer) [inline test/pure-component-test.output.js]: I still like the const + arrow function mode. If you wanna go crazy, how about making it an option for the hipster people? :)

**Comment by cpojer** [comment]
This is a good start! The tests should be fixed on master now.

* Maybe take a look at the pure-render-mixin and class transforms in this file for the general structure. I think this codemod should be defensive and only inline actually pure components. It could print a `console.warn` that says that it skipped a class (`class.js` has a nice copy-paste-able warning thing).
* Move everything into `module.exports` to make it actually work
* Maybe add this: https://github.com/reactjs/react-codemod/blob/master/transforms/pure-render-mixin.js#L168 for safety!

File: transforms/pure-component.js (hunk 4 above)
Comment (thejameskyle) [inline transforms/pure-component.js]: I started refactoring and then tried to write tests and the tests didn't work and I didn't have time to look into it earlier.

**Comment by fkling** [comment]
> It could print a console.warn that says that it skipped a class (class.js has a nice copy-paste-able warning thing).

There is also the `stats` function. It's basically a keyed counter. If you do a `dry` run, it prints how often `stats` was called with that argument. E.g. if a class with more than a render function is encountered, you could call `api.stats('complex class')` and it would print something like `complex class: 4` in the dry run.

Useful for collecting statistics about your codebase / codemod.

**Comment by thejameskyle** [comment]
Threw together a Babel plugin as well. Thought it'd be a useful optimization to run in a production build https://github.com/thejameskyle/babel-plugin-react-pure-components

File: transforms/pure-component.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+  pureClasses.replaceWith(p => {
+    const name = p.node.id.name;
+    const renderMethod = p.value.body.body.filter(isRenderMethod)[0];
+    const renderBody = renderMethod.value.body;
+
+    replaceThisProps(renderBody);
+
+    if (useArrows) {
+      return buildPureComponentArrowFunction(name, renderBody);
```
Comment (cpojer) [inline transforms/pure-component.js]: this variant should probably take the `argument` of the `ReturnStatement` if the length of `renderBody.body` is one and it only contains a return statement :)

**Comment by cpojer** [comment]
Yay! Would you mind flattening your commits into a single one before I merge? :)

**Comment by cpojer** [comment]
Dammit, I missed this because github doesn't notify me about force pushes to PRs.

**Comment by thejameskyle** [comment]
Sorry, I knew that, should've pinged you.

**Pull request merged**